sources/osgShadow/ShadowMapAtlas.js

```diff
@@ -0,0 +1,600 @@
+            this._shadowMaps[ i ].setRotateOffset( value );
+        }
+    },
+
+    setShadowedScene: function ( shadowedScene ) {
+
+        ShadowTechnique.prototype.setShadowedScene.call( this, shadowedScene );
+        this._receivingStateset = this._shadowedScene.getReceivingStateSet();
+
+        for ( var i = 0, l = this._shadowMaps.length; i < l; i++ ) {
+            this._shadowMaps[ i ].setShadowedScene( shadowedScene );
+        }
+
+    },
+
+    setTexturePrecision: function ( value ) {
+
+        for ( var i = 0, l = this._shadowMaps.length; i < l; i++ ) {
+            this._shadowMaps[ i ].setTexturePrecision( value );
+        }
+
+    },
+
+    getTexturePrecision: function ( numShadow ) {
+
+        if ( numShadow !== undefined ) {
+            return this._shadowMaps[ numShadow ].getTexturePrecision();
+        } else if ( this._shadowMaps.length !== 1 ) {
+            return this._shadowMaps[ 0 ].getTexturePrecision();
+        }
+
+    },
+
+    setShadowMapSize: function ( mapSize, lightNum ) {
+
+
+        this._shadowMapSize = mapSize;
+        this._numShadowWidth = this._textureSize / this._shadowMapSize;
+        this._numShadowHeight = this._textureSize / this._shadowMapSize;
+
+        if ( !lightNum ) {
+            for ( var i = 0, l = this._shadowMaps.length; i < l; i++ ) {
+                this._shadowMaps[ i ].setTextureSize( mapSize );
+            }
+        } else {
+            this._shadowMaps[ lightNum ].setTextureSize( mapSize );
+        }
+
+    },
+
+    setTextureSize: function ( mapSize ) {
+
+        if ( mapSize === this._textureSize ) return;
+
+        //this._textureSize = mapSize;
+        //this._textureSize = settings.atlasSize;
+        this._shadowMapSize = mapSize;
+
+
+        this._numShadowWidth = this._textureSize / this._shadowMapSize;
+        this._numShadowHeight = this._textureSize / this._shadowMapSize;
+
+        this.dirty();
+    },
+
+    setAlgorithm: function ( algo, lightNum ) {
+
+        if ( !lightNum ) {
+            for ( var i = 0, l = this._shadowMaps.length; i < l; i++ ) {
+                this._shadowMaps[ i ].setAlgorithm( algo );
+            }
+        } else {
+            this._shadowMaps[ lightNum ].setAlgorithm( algo );
+        }
+
+    },
+
+    getAlgorithm: function ( numShadow ) {
+
+        if ( numShadow !== undefined ) {
+            return this._shadowMaps[ numShadow ].getAlgorithm();
+        } else if ( this._shadowMaps.length !== 1 ) {
+            return this._shadowMaps[ 0 ].getAlgorithm();
+        }
+
+    },
+
+    getShadowMap: function ( lightNum ) {
+        return this._shadowMaps[ lightNum ];
+    },
+
+    addLight: function ( light, settings, customMapSize ) {
+
+        if ( !light )
+            return -1;
+
+        var lightNum = this._lights.indexOf( light );
+        if ( lightNum !== -1 )
+            return lightNum;
+
+        this._lights.push( light );
+        lightNum = this._lights.length;
```

Comment:
num is confusing and there code that can't work. A simplified version could look like this
```

    addLight: function ( light ) {

        if ( !light ) return undefined;

        var lightIndex = this._lights.indexOf( light );
        if ( lightIndex !== -1 ) return undefined;

        if ( this._lights.length === ( this._numShadowWidth * this._numShadowHeight - 1 ) ) {
            notify.warn( 'can\'t allocate shadow for light ' + light.getLightNumber() + ' ShadowAtlas already full ' );
            return undefined;
        }

        this._lights.push( light );

        var shadowMap = new ShadowMap( this._shadowSettings, this._texture );
        this._shadowMaps.push( shadowMap );

        var lighCount = this._lights.length;

        var y = this._shadowMapSize * ( lighCount % ( this._numShadowWidth ) );
        var x = this._shadowMapSize * ( Math.floor( lighCount / ( this._numShadowHeight ) ) );

        shadowMap.setShadowedScene( this._shadowedScene );
        this._viewportDimension.push( vec4.fromValues( x, y, this._shadowMapSize, this._shadowMapSize ) );

        return shadowMap;
    },


```